File: trs-sdk-java/tests/InteropNodeFlow.java
package dev.trs.sdk;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class InteropNodeFlow {
    public static void main(String[] args) throws Exception {
        if (args.length != 2) {
            throw new IllegalArgumentException("usage: InteropNodeFlow <base_url> <output_path>");
        }
        String baseUrl = args[0];
        Path outputPath = Path.of(args[1]);

        TRSClient client = new TRSClient(baseUrl, Duration.ofSeconds(5));
        HealthStatus health = client.health();

        Map<String, Object> observation = new LinkedHashMap<>();
        observation.put("id", "java-obs-1");
        observation.put("type", "Observation");
        observation.put("author", "java-sdk");
        observation.put("timestamp", "2026-08-05T15:05:02+00:00");
        observation.put("schema", "trs.observation.v1");
        observation.put("payload", Map.of("subject", "interop", "value", 2));
        observation.put("causes", List.of());
        observation.put("authorization", List.of());
        observation.put("signature", "sig:java-obs-1");
        SubmitResult submit = client.submit(observation);

        Map<String, Object> intention = new LinkedHashMap<>();
        intention.put("id", "java-int-1");
        intention.put("type", "Intention");
        intention.put("author", "java-sdk");
        intention.put("timestamp", "2026-08-05T15:05:03+00:00");
        intention.put("schema", "trs.intention.v1");
        intention.put("payload", Map.of("goal", "interop-check", "horizon", "cycle0012"));
        intention.put("causes", List.of("java-obs-1"));
        intention.put("authorization", List.of());
        intention.put("signature", "sig:java-int-1");

        SyncResult sync = client.sync(List.of(intention));
        List<Map<String, Object>> records = client.query(Map.of("author", "java-sdk"));
        Map<String, Object> replay = client.replay();

        List<?> unresolved = List.of();
        Object coordination = replay.get("coordination");
        if (coordination instanceof Map<?, ?> map) {
            Object unresolvedRaw = map.get("unresolved_intentions");
            if (unresolvedRaw instanceof List<?> values) {
                unresolved = values;
            }
        }

        Map<String, Object> summary = new LinkedHashMap<>();
        summary.put("sdk", "java");
        summary.put("base_url", baseUrl);
        summary.put(
                "health",
                Map.of("status", health.status(), "runtime", health.runtime(), "node", health.node()));
        summary.put(
                "submit",
                Map.of(
                        "accepted", submit.accepted(),
                        "record_id", submit.recordId(),
                        "errors", submit.errors()));
        summary.put(
                "sync",
                Map.of(
                        "accepted_count", sync.acceptedCount(),
                        "rejected_count", sync.rejectedCount(),
                        "appended_ids", sync.appendedIds(),
                        "rejected_errors", sync.rejectedErrors()));
        summary.put("query_author_count", records.size());
        summary.put("replay_unresolved_intentions", unresolved);

        String json = Jsons.stringify(summary);
        Files.writeString(outputPath, json, StandardCharsets.UTF_8);
        System.out.println(json);
    }
}

File: trs-sdk-rust/examples/interop_node_flow.rs
use std::env;
use std::fs;
use std::time::Duration;

use serde_json::json;
use trs_sdk_rust::TrsClient;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut args = env::args();
    let _program = args.next();
    let base_url = args
        .next()
        .ok_or_else(|| "usage: cargo run --example interop_node_flow -- <base_url> <output_path>")?;
    let output_path = args
        .next()
        .ok_or_else(|| "usage: cargo run --example interop_node_flow -- <base_url> <output_path>")?;

    let client = TrsClient::new(base_url.clone(), Duration::from_secs(5));
    let health = client.health()?;

    let observation = json!({
        "id": "rust-obs-1",
        "type": "Observation",
        "author": "rust-sdk",
        "timestamp": "2026-08-05T15:05:00+00:00",
        "schema": "trs.observation.v1",
        "payload": {"subject": "interop", "value": 1},
        "causes": [],
        "authorization": [],
        "signature": "sig:rust-obs-1"
    });
    let submit_result = client.submit(observation)?;

    let intention = json!({
        "id": "rust-int-1",
        "type": "Intention",
        "author": "rust-sdk",
        "timestamp": "2026-08-05T15:05:01+00:00",
        "schema": "trs.intention.v1",
        "payload": {"goal": "interop-check", "horizon": "cycle0012"},
        "causes": ["rust-obs-1"],
        "authorization": [],
        "signature": "sig:rust-int-1"
    });
    let sync_result = client.sync(vec![intention])?;
    let queried = client.query(json!({"author": "rust-sdk"}))?;
    let replay = client.replay()?;
    let unresolved_intentions = replay
        .get("coordination")
        .and_then(|coord| coord.get("unresolved_intentions"))
        .and_then(|raw| raw.as_array())
        .map(|values| values.iter().filter_map(|v| v.as_str().map(ToOwned::to_owned)).collect::<Vec<String>>())
        .unwrap_or_default();

    let summary = json!({
        "sdk": "rust",
        "base_url": base_url,
        "health": {
            "status": health.status,
            "runtime": health.runtime,
            "node": health.node
        },
        "submit": {
            "accepted": submit_result.accepted,
            "record_id": submit_result.record_id,
            "errors": submit_result.errors
        },
        "sync": {
            "accepted_count": sync_result.accepted_count,
            "rejected_count": sync_result.rejected_count,
            "appended_ids": sync_result.appended_ids,
            "rejected_errors": sync_result.rejected_errors
        },
        "query_author_count": queried.len(),
        "replay_unresolved_intentions": unresolved_intentions
    });

    fs::write(output_path, serde_json::to_string_pretty(&summary)?)?;
    println!("{}", serde_json::to_string_pretty(&summary)?);
    Ok(())
}
